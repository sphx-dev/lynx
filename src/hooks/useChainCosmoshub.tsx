import {
  useAccount,
  useConnect,
  useDisconnect,
  // useStargateClient,
  WalletType,
} from "graz";
import { Keplr } from "@keplr-wallet/types";
import { getChain } from "../config";

// export const useChainCosmoshub = () => useChain("cosmoshub");
export const useChainCosmoshub = () => {
  const { connectAsync, status: connectionStatus, error } = useConnect();
  const {
    data: account,
    isConnected,
    status: accountStatus,
    reconnect,
  } = useAccount();
  const { disconnect } = useDisconnect();

  // TODO: use client to get sequence if needed for order creation
  // const { data: client } = useStargateClient();

  // (async () => {
  //   console.log("ACCOUNT", account);
  //   console.log("CLIENT", client);
  //   console.log("CLIENT seq", client?.getSequence);
  //   console.log(
  //     "CLIENT seq addr",
  //     await client?.getSequence(account?.bech32Address)
  //   );
  // })();

  if (error) {
    console.log("useConnect error", error);
  }

  return {
    addChain: async () => {
      try {
        const keplr: Keplr = window.keplr;
        await keplr.experimentalSuggestChain(getChain());
      } catch (e) {
        console.log("addChain ERR -------->", e);
      }
    },
    connect: async () => {
      try {
        await connectAsync({
          chainId: getChain().chainId,
          walletType: WalletType.KEPLR,
        });
      } catch (e) {
        console.log("-------->", e);
      }
    },
    reconnect,
    disconnect,
    connectionStatus,
    accountStatus,
    account,
    address: account?.bech32Address,
    isConnected,
  };
};
