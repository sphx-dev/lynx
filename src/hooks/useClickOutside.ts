import { useEffect, MutableRefObject } from "react";

const useClickOutside = (
  ref:
    | MutableRefObject<HTMLElement | null>
    | MutableRefObject<HTMLElement | null>[],
  handler: () => void
): void => {
  const refs = Array.isArray(ref) ? ref : [ref];
  const shouldTrigger = (event: MouseEvent) =>
    refs.every(
      ref => ref.current && !ref.current.contains(event.target as Node)
    );
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (shouldTrigger(event)) {
        handler();
      }
    };

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [handler]);
};

export default useClickOutside;
