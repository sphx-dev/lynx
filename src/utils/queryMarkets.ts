import { getChain } from "../config";
import { createRPCQueryClient } from "../../proto-codecs/codegen/sphx/rpc.query";
import { sphx } from "../../proto-codecs";
import { MarketStatus } from "../../proto-codecs/codegen/sphx/order/market";

export const getAllMarkets = async () => {
  const queryClient = await createRPCQueryClient({
    rpcEndpoint: getChain().rpc,
  });

  const response = await queryClient.sphx.order.markets({});

  return response;
};

export const createMarket = async () => {
  // TODO: build a message like this
  const message = sphx.order.MessageComposer.withTypeUrl.registerMarket({
    authority: "TODO: set authority here",
    market: {
      id: BigInt(1),
      ticker: "BTCUSDT.P",
      status: MarketStatus.STATUS_ACTIVE,
      baseAsset: "BTC",
      quoteAsset: "USDT",
    },
  });
  console.log("createMarket msg:", message);
  // TODO: get signing client
  // TODO: sign and broadcast
};
