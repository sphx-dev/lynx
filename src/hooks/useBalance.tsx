import { getBalance } from "../utils/getBalance";
import { formatCoin } from "../utils/formatCoin";
import { createStore } from "zustand";
import { useEffect, useState } from "react";

export const useBalance = (
  address: string | undefined | null,
  denom = "uusdc"
) => {
  const store = balancesStore.getState();
  const [balance, setBalance] = useState<Balance | null>(null);

  useEffect(() => {
    (async () => {
      if (!address) return;

      const freshBalance = await store.getBalance(address, denom);
      setBalance(freshBalance);
    })();
  }, [address, denom, store, store.balances, setBalance]);

  return {
    amount: balance?.amount,
    formatedAmount: formatCoin(Number(balance?.amount) / 1e6 || 0),
    denom: balance?.denom,
    markBalanceAsStale: () => markBalanceAsStale(address || "", denom),
  };
};

export const useBalances = (balances: { address: string; denom: string }[]) => {
  const store = balancesStore.getState();
  const [balanceList, setBalanceList] = useState<(Balance | null)[]>([]);

  useEffect(() => {
    (async () => {
      if (!balances.length) return;
      let freshBalances = [];
      for (let i = 0; i < balances.length; i++) {
        freshBalances[i] = await store.getBalance(
          balances[i].address,
          balances[i].denom
        );
      }
      setBalanceList(freshBalances);
    })();
  }, [balances, store]);

  return {
    balances: balanceList,
  };
};

type Balance = {
  address: string;
  amount: number;
  denom: string;
  isStale: boolean;
};

export const markBalanceAsStale = (address: string, denom: string) => {
  const store = balancesStore.getState();
  store.markBalanceStale(address, denom);
};

type BalancesStore = {
  balances: Balance[];
  getBalance: (address: string, denom: string) => Promise<Balance | null>;
  setBalance: (balance: Balance) => void;
  markBalanceStale: (address: string, denom: string) => void;
};

const balancesStore = createStore<BalancesStore>((set, get) => ({
  balances: [] as Balance[],
  getBalance: async (address: string, denom: string) => {
    if (!address || !denom) return null;
    const balance = get().balances.find(
      (balance: Balance) =>
        balance.address === address && balance.denom === denom
    );
    if (!balance || balance.isStale) {
      const currentBalance = await getBalance(address, denom);

      get().setBalance({
        address,
        amount: parseInt(currentBalance.amount),
        denom,
        isStale: false,
      });
      return {
        address,
        amount: parseInt(currentBalance.amount),
        denom,
        isStale: false,
      } as Balance;
    } else {
      return { amount: balance.amount, denom: balance.denom } as Balance;
    }
  },
  setBalance: ({
    address,
    amount,
    denom,
  }: {
    address: string;
    amount: number;
    denom: string;
  }) => {
    set(state => {
      const balanceExists = state.balances.find(
        b => b.address === address && b.denom === denom
      );
      return {
        balances: balanceExists
          ? state.balances.map((b: Balance) =>
              b.address === address && b.denom === denom
                ? { ...b, address, amount, denom, isStale: false }
                : b
            )
          : [
              ...state.balances,
              {
                address,
                amount,
                denom,
                isStale: false,
              },
            ],
      };
    });
  },
  markBalanceStale: (address: string, denom: string) => {
    set(state => ({
      balances: state.balances.map((balance: Balance) =>
        balance.address === address && balance.denom === denom
          ? { ...balance, isStale: true }
          : balance
      ),
    }));
  },
}));
