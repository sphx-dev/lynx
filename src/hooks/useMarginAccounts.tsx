import {
  createMarginAccount,
  getAccountsByOwner,
} from "../utils/queryMarginAccounts";
import { MarginAccountInfo } from "proto-codecs/codegen/sphx/marginacc/margin_account";
import { create } from "zustand";
import { useCallback, useEffect } from "react";
import { errorAlert, successAlert } from "@/utils/alerts";
import { useMutation, useQuery, useQueryClient } from "react-query";
import { useTranslation } from "react-i18next";

type MarginAccountState = {
  marginAccounts: MarginAccountInfo[];
  selectedIndex: number;

  setMarginAccounts: (marginAccounts: MarginAccountInfo[]) => void;
  setSelectedIndex: (selectedIndex: number) => void;

  getSelectedAccount: () => MarginAccountInfo | undefined;
  getSelectedAddress: () => string | undefined;
  getSelectedIndex: () => number;
};

export const useMarginAccount = (owner: string | null | undefined) => {
  const { t } = useTranslation();
  const store = useMarginAccountStore();
  const { setMarginAccounts } = store;

  const queryClient = useQueryClient();

  // Fetch margin accounts
  const { data, isLoading, isError, isIdle, isSuccess, status } = useQuery(
    ["marginAccounts", owner],
    () => {
      return getAccountsByOwner(owner!).then(response => {
        return response.marginAccounts;
      });
    },
    {
      enabled: !!owner,
      staleTime: 1000 * 60 * 60,
    }
  );
  useEffect(() => {
    if (data) {
      setMarginAccounts(data);
    }
  }, [data, setMarginAccounts]);

  // Mutation for creating margin account
  const mutation = useMutation(createMarginAccount);
  const createAccount = useCallback(() => {
    const ids = store.marginAccounts.map(ma => ma?.id?.number ?? 0);
    const nextMarginAccountNumber = Math.max(...ids, 0) + 1;
    return mutation.mutateAsync(
      {
        address: owner!,
        number: nextMarginAccountNumber,
        memo: t("createMarginAccount"),
      },
      {
        onSettled: () => {
          queryClient.invalidateQueries(["marginAccounts", owner]);
        },
        onSuccess: response => {
          if (response.code === 0) {
            successAlert("Margin account created successfully");
          } else {
            errorAlert(getTextByCode(response.code));
          }
        },
        onError: error => {
          errorAlert(getTextFromError((error as Error).message));
        },
      }
    );
  }, [t, owner, mutation, store.marginAccounts, queryClient]);

  return {
    ...store,
    status,
    isIdle,
    isLoading,
    isError,
    isSuccess,
    selectedAccount: store.getSelectedAccount(),
    selectedAddress: store.getSelectedAddress(),
    selectedIndex: store.getSelectedIndex(),
    createAccount,
  };
};

export const useMarginAccountStore = create<MarginAccountState>((set, get) => ({
  marginAccounts: [],
  selectedIndex: 0,

  // Setters
  setMarginAccounts: (marginAccounts: MarginAccountInfo[]) => {
    set({
      marginAccounts,
    });
  },
  setSelectedIndex: (selectedIndex: number) => {
    set({
      selectedIndex,
    });
  },

  // Getters
  getSelectedAccount: () => {
    return get().marginAccounts[get().selectedIndex];
  },
  getSelectedAddress: () => {
    return get().marginAccounts[get().selectedIndex]?.address;
  },
  getSelectedIndex: () => {
    return get().selectedIndex;
  },
}));

//TODO: move this to translations
const getTextByCode = (code: number) => {
  switch (code) {
    case 9:
      return "Margin account already exists";
    default:
      return "Error creating margin account. Please try again";
  }
};

const getTextFromError = (message: string) => {
  if (
    message.includes("Send some tokens there before trying to query sequence")
  ) {
    return "Seems like youâ€™re using new wallet. You must have at least 0.01 USDC for transaction fee";
  }

  return "Something went wrong. Please try again";
};
