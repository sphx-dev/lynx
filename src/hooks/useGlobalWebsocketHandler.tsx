import { useEffect, useRef, useCallback } from "react";
import { useQueryClient } from "react-query";
import { useChainCosmoshub } from "./useChainCosmoshub";
import {
  CosmosTransactionEvent,
  CosmosFlattenedEvent,
  JsonRpcResultMessage,
  useWebsocket,
} from "./useWebsocket";
import { useMarginAccount } from "./useMarginAccounts";

export function useGlobalWebsocketHandler() {
  const { address } = useChainCosmoshub();
  const { selectedAddress } = useMarginAccount(address);
  const queryClient = useQueryClient();
  const messageHandlerRef = useRef<(message: JsonRpcResultMessage) => void>();

  const handleWebSocketMessage = useCallback(
    (message: JsonRpcResultMessage) => {
      // Ensure the address is available
      if (!address) {
        return;
      }

      if (!selectedAddress) {
        return;
      }

      const chainEvents =
        message?.result?.data?.value?.TxResult?.result?.events;
      const events: {
        [eventName: string]: CosmosFlattenedEvent;
      } = message?.result.events;

      // Ensure the chain events are available
      if (!chainEvents) {
        return;
      }

      // Ensure message sender is the current address or MarginAcount address
      console.log("WS_message", "message.sender", events["message.sender"]);
      if (
        !events["message.sender"].includes(address) &&
        !events["message.sender"].includes(selectedAddress)
      ) {
        return;
      }

      const chainEventsNames = chainEvents.map((event) => event.type);
      const chainEventsMap = new Set(chainEventsNames);

      // ACCOUNT EVENTS
      if (chainEventsMap.has("create_margin_account")) {
        console.log("WS_message", "create_margin_account event");
        queryClient.invalidateQueries(["marginAccounts"]);
      }
      if (chainEventsMap.has("withdraw")) {
        console.log("WS_message", "withdraw event");
      }

      // ORDER EVENTS
      // "register_market"
      if (chainEventsMap.has("register_market")) {
        console.log("WS_message", "register_market event");
        queryClient.invalidateQueries(["markets"]);
      }
      // "place_order"
      if (chainEventsMap.has("place_order")) {
        console.log("WS_message", "place_order event");
        queryClient.invalidateQueries(["orders"]);
      }
      // "cancel_order"
      if (chainEventsMap.has("cancel_order")) {
        console.log("WS_message", "cancel_order event");
        queryClient.invalidateQueries(["orders"]);
      }

      // "new_position" and "modify_position"
      if (chainEventsMap.has("new_position")) {
        console.log("WS_message", "new_position");
        queryClient.invalidateQueries(["positions"]);
      }
      if (chainEventsMap.has("modify_position")) {
        console.log("WS_message", "modify_position");
        queryClient.invalidateQueries(["positions"]);
      }

      // TRADE EVENTS
      // "transfer"
      if (chainEventsMap.has("transfer")) {
        const addresses = getAddresesFromTransferEvent(chainEvents);
        console.log(
          "WS_message",
          "transfer event",
          addresses,
          addresses.includes(address)
        );
        if (addresses.includes(address)) {
          queryClient.invalidateQueries(["balance"]);
        }
      }
    },
    [address, queryClient, selectedAddress]
  );

  // Persist the current callback in a ref to avoid resubscribing
  useEffect(() => {
    messageHandlerRef.current = handleWebSocketMessage;
  }, [handleWebSocketMessage]);

  // Subscribe to WebSocket messages using the singleton pattern
  useWebsocket((message) => {
    if (messageHandlerRef.current) {
      messageHandlerRef.current(message);
    }
  });
}

function getAddresesFromTransferEvent(events: CosmosTransactionEvent[]) {
  const transferEvents = events.filter((ev) => ev.type === "transfer");
  const attributes = transferEvents.map((ev) => ev.attributes).flat();
  const addresses = attributes
    .filter((attr) => attr.key === "recipient" || attr.key === "sender")
    .map((attr) => attr.value);
  return addresses;
}
