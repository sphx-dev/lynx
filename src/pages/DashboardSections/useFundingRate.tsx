import config from "@/config";
import { useChainCosmoshub } from "@/hooks/useChainCosmoshub";
import { useMarginAccount } from "@/hooks/useMarginAccounts";
import { useMarkets } from "@/hooks/useMarkets";
import axios from "axios";
import { useQuery } from "react-query";

const BASE_API = config.VITE_API_URL;

export const useFundingRate = (from: number, to: number) => {
  console.log("[useFundingRate]", from, to);
  const { address, isConnected } = useChainCosmoshub();
  const { selectedAddress } = useMarginAccount(address);
  const { selectedMarket } = useMarkets();
  const ticker = selectedMarket?.ticker;

  const { data } = useQuery({
    queryKey: ["funding", address, ticker, from, to],
    queryFn: async () => {
      console.log("[useFundingRate] Fetching funding data");

      const data = await fetchFundingRateLogs(
        selectedAddress!,
        ticker!,
        from,
        to
      );
      console.log(data);
      return data;
    },
    enabled: isConnected && !!address && !!ticker,
    staleTime: 1000 * 60 * 5,
  });

  return data;
};

export async function fetchFundingRateLogs(
  account_id: string,
  symbol: string,
  from: number,
  to: number
) {
  const URL = `${BASE_API}/other/funding_rate_logs?account_id=${account_id}&symbol=${symbol}&from=${from}&to=${to}`;

  try {
    axios.defaults.withCredentials = true;

    const response = await axios.get(URL, {
      withCredentials: true,
      headers: {
        "X-ACCOUNT-ID": account_id,
      },
    });
    return response.data;
  } catch (error) {
    console.error(error);
    return [];
  }
}
