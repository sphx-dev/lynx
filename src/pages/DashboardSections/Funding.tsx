import { useCallback, useEffect, useState } from "react";
import { useChainCosmoshub } from "@/hooks/useChainCosmoshub";
import { useMarginAccount } from "@/hooks/useMarginAccounts";
import { useMarkets } from "@/hooks/useMarkets";
import DatePicker from "@/components/ReactDatepickerCustom/DatePicker";
import { Button, Divider, Table } from "@/components";
import { useTranslation } from "react-i18next";
import dayjs from "dayjs";
import styled from "styled-components";
import SymbolSelect from "@/components/SymbolSelect/SymbolSelect";
import { useFundingRateColumns } from "./useFundingRateColumns";
import { fetchFundingRateLogs } from "./useFundingRate";
import { Title } from "./styled";
import { Pagination } from "@/components/Pagination";

export const Funding = () => {
  const { t } = useTranslation();
  const { address, isConnected } = useChainCosmoshub();
  const { selectedAddress } = useMarginAccount(address);
  const { selectedMarket } = useMarkets();

  const [fromDate, setFromDate] = useState<Date | null>(
    dayjs(Date.now()).subtract(1, "month").toDate()
  );
  const [toDate, setToDate] = useState<Date | null>(new Date());

  const [data, setData] = useState(null);
  const onSubmit = useCallback(async () => {
    if (
      !isConnected ||
      !selectedAddress ||
      !selectedMarket?.ticker ||
      !fromDate ||
      !toDate
    ) {
      return;
    }

    const data = await fetchFundingRateLogs(
      selectedAddress!,
      selectedMarket?.ticker,
      Math.floor(fromDate.getTime() / 1000),
      Math.floor(toDate.getTime() + (24 * 60) / 1000)
    );
    setData(data);
  }, [isConnected, selectedAddress, selectedMarket, fromDate, toDate]);

  useEffect(() => {
    if (!Boolean(selectedAddress)) {
      setData(null);
    } else {
      onSubmit();
    }
  }, [selectedAddress, setData, onSubmit]);

  const loadButtonDisabled =
    !isConnected ||
    !selectedAddress ||
    !selectedMarket?.ticker ||
    !fromDate ||
    !toDate;

  return (
    <div>
      <Title>{t("fundingRateLogs")}</Title>

      <div
        style={{
          display: "flex",
          gap: "18px",
          alignItems: "end",
          minHeight: "60px",
        }}
      >
        <div
          style={{
            display: "flex",
            alignItems: "end",
            minHeight: "50px",
          }}
        >
          <SymbolSelect />
        </div>
        <DatePickerContainer>
          <DatePickerLabel>{t("From")}:</DatePickerLabel>
          <DatePicker
            onChange={date => {
              // setTimeout fixes interblocking issue with selected
              setTimeout(() => setFromDate(date), 100);
            }}
            selected={fromDate}
            dateFormat="yyyy/MM/dd"
          />
        </DatePickerContainer>
        <DatePickerContainer>
          <DatePickerLabel>{t("To")}:</DatePickerLabel>
          <DatePicker
            onChange={date => {
              // setTimeout fixes interblocking issue with selected
              setTimeout(() => setToDate(date), 100);
            }}
            selected={toDate}
            dateFormat="yyyy/MM/dd"
          />
        </DatePickerContainer>
        <Button onClick={onSubmit} disabled={loadButtonDisabled}>
          Load
        </Button>
      </div>

      <Divider style={{ margin: "1rem" }} />

      {!data ? (
        <div
          style={{
            display: "flex",
            justifyContent: "center",
            alignContent: "center",
            height: "200px",
          }}
        >
          <div
            style={{
              display: "flex",
              flexDirection: "column",
              justifyContent: "center",
              alignContent: "center",
              gap: "1rem",
            }}
          >
            {!isConnected ? (
              t("conectWalletToViewFundingRateLogs")
            ) : (
              <>
                {t("selectRageToLoadData")}
                <Button onClick={onSubmit}>Load</Button>
              </>
            )}
          </div>
        </div>
      ) : (
        <FundingRateTable data={data} />
      )}
    </div>
  );
};

const PAGE_SIZE = 15;
const FundingRateTable = ({ data }: { data: any[] }) => {
  const columns = useFundingRateColumns();

  const [page, setPage] = useState(0);
  const pageData = data.slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE);

  return (
    <>
      <div style={{ minHeight: "530px" }}>
        <Table columns={columns} data={pageData}></Table>
      </div>
      <Pagination
        totalItems={data.length}
        page={page}
        setPage={setPage}
        pageSize={PAGE_SIZE}
      />
    </>
  );
};

const DatePickerContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 2px;
  align-items: flex-start;
`;

const DatePickerLabel = styled.label`
  font-family: "Poppins-Regular";
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  color: #ffffff;
  text-align: left;
`;
