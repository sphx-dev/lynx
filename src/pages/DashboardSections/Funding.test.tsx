import { render, screen, waitFor } from "@testing-library/react";
import { Funding } from "./Funding";
import { useChainCosmoshub } from "@/hooks/useChainCosmoshub";
import { useMarginAccount } from "@/hooks/useMarginAccounts";
// import { useMarkets } from "@/hooks/useMarkets";
import { getAllMarkets } from "@/utils/queryMarkets";
import { fetchFundingRateLogs } from "./useFundingRate";
import { Mock, vi } from "vitest";
import { AppWrapper } from "@/testing/AppWrapper";

import { act } from "react";
// import { act } from "@testing-library/react";

vi.mock("@/hooks/useChainCosmoshub", () => ({
  useChainCosmoshub: vi.fn(),
}));

vi.mock("@/hooks/useMarginAccounts", () => ({
  useMarginAccount: vi.fn(),
}));

vi.mock("@/utils/queryMarkets", () => ({
  getAllMarkets: vi.fn(),
}));

vi.mock("./useFundingRate", () => ({
  fetchFundingRateLogs: vi.fn(),
}));

describe("Funding", () => {
  const address = "test-address";
  const selectedAddress = "selected-address";
  const markets = [
    {
      id: 1n,
      ticker: "BTCUSDC.P",
      status: 1,
      baseAsset: "BTC",
      quoteAsset: "USDC",
    },
    {
      id: 2n,
      ticker: "ETHUSDC.P",
      status: 1,
      baseAsset: "ETH",
    },
  ];
  const isConnected = true;

  beforeEach(() => {
    (useChainCosmoshub as Mock).mockReturnValue({ address, isConnected });
    (useMarginAccount as Mock).mockReturnValue({ selectedAddress });
    (getAllMarkets as Mock).mockReturnValue({
      markets,
    });
    (fetchFundingRateLogs as Mock).mockResolvedValue([]);
    // screen.getByTestId("connect-test-button").click();
  });

  it("renders the Funding component", () => {
    act(() => {
      render(<Funding />, { wrapper: AppWrapper });
    });

    expect(screen.getByText("Funding Rate Logs")).toBeInTheDocument();
  });

  it("disables the Load button when required fields are missing", () => {
    (useChainCosmoshub as Mock).mockReturnValue({
      address: null,
      isConnected: false,
    });
    act(() => {
      render(<Funding />, { wrapper: AppWrapper });
    });
    expect(screen.getByText("Load")).toBeDisabled();
  });

  it("calls fetchFundingRateLogs on Load button click", async () => {
    (useChainCosmoshub as Mock).mockReturnValue({
      address: "sphx1test",
      isConnected: true,
    });

    (fetchFundingRateLogs as Mock).mockReturnValue([
      {
        id: 1,
        ticker: "BTCUSDC.P",
        size: 1.5,
        funding_rate: 0.0001,
        funding_amount: 0.00015,
        accumulated_funding: 0.00015,
        side: "buy",
        timestamp: "2021-09-01T12:35:00Z",
      },
    ]);

    act(() => {
      render(<Funding />, {
        wrapper: AppWrapper,
      });
    });

    expect(fetchFundingRateLogs).toHaveBeenCalledWith(
      selectedAddress,
      "BTCUSDC.P",
      expect.any(Number),
      expect.any(Number)
    );
  });

  it("displays a message when not connected", () => {
    (useChainCosmoshub as Mock).mockReturnValue({
      address,
      isConnected: false,
    });
    act(() => {
      render(<Funding />, { wrapper: AppWrapper });
    });
    expect(
      screen.getByText(
        "Connect wallet and select a margin account to view funding rate logs"
      )
    ).toBeInTheDocument();
  });

  it("displays the FundingRateTable when data is available", async () => {
    (useChainCosmoshub as Mock).mockReturnValue({
      address: "sphx1test",
      isConnected: true,
    });

    (fetchFundingRateLogs as Mock).mockReturnValue([
      {
        id: 1,
        ticker: "BTCUSDC.P",
        size: 1.5,
        funding_rate: 0.0001,
        funding_amount: 0.00015,
        accumulated_funding: 0.00015,
        side: "buy",
        timestamp: "2021-09-01T12:35:00Z",
      },
    ]);

    let renderResult: any;
    act(() => {
      renderResult = render(<Funding />, {
        wrapper: AppWrapper,
      });
    });
    const { container, getAllByText } = renderResult;

    await waitFor(() => {
      const elements = getAllByText("BTCUSD.P");
      // Two elements containing "BTCUSD.P" should be rendered
      // 1. SymbolSelector
      // 2. Table row
      expect(elements.length).toBe(2);
    });

    expect(getAllByText("BTCUSD.P")).toBeTruthy();
    expect(getAllByText("BTCUSD.P")[0]).toBeInTheDocument();

    expect(fetchFundingRateLogs).toHaveBeenCalledWith(
      selectedAddress,
      "BTCUSDC.P",
      expect.any(Number),
      expect.any(Number)
    );

    await waitFor(() =>
      expect(container.querySelector("tbody")).toBeInTheDocument()
    );

    expect(container.querySelector("table")?.innerHTML).toMatchSnapshot();
  });
});
