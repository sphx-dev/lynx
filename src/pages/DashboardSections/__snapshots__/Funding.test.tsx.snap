// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Funding > displays the FundingRateTable when data is available 1`] = `"<thead><tr><th><div class="sc-dpBQxM hlYqZS">Market</div></th><th><div class="sc-dpBQxM hlYqZS">Qty</div></th><th><div class="sc-dpBQxM hlYqZS">Funding Rate</div></th><th><div class="sc-dpBQxM hlYqZS">Funding Amount</div></th><th><div class="sc-dpBQxM hlYqZS">Accumulated Funding</div></th><th><div class="sc-dpBQxM hlYqZS">Side</div></th><th><div class="sc-dpBQxM hlYqZS">Timestamp</div></th></tr></thead><tbody><tr class="sc-keTIit hZSLjO"><td><p align="left" class="sc-blHHSb dsFwju text-sm text-tertiary">BTCUSD.P</p></td><td><p align="left" class="sc-blHHSb dsFwju text-sm text-tertiary">1.5</p></td><td><p align="left" class="sc-blHHSb dsFwju text-sm text-tertiary">0.0001 %</p></td><td><p align="left" class="sc-blHHSb dsFwju text-sm text-tertiary">0.00015 USD</p></td><td><p align="left" class="sc-blHHSb dsFwju text-sm text-tertiary">0.00015 USD</p></td><td><p align="left" class="sc-blHHSb dsFwju text-sm text-buy"><span style="text-transform: capitalize;">buy</span></p></td><td><p align="left" class="sc-blHHSb dsFwju text-sm text-tertiary">2021-09-01 12:35:00</p></td></tr></tbody><tfoot><tr><th></th><th></th><th></th><th></th><th></th><th></th><th></th></tr></tfoot>"`;
